lib/src/package_analysis/common.dart

```diff
@@ -1,12 +1,204 @@
-  AnalysisContextCollection get analysisContextCollection;
+import 'package:analyzer/dart/analysis/session.dart';
+import 'package:analyzer/file_system/file_system.dart' as resource;
+import 'package:pana/pana.dart';
+import 'package:path/path.dart' as path;
+import 'package:pub_semver/pub_semver.dart';
+import 'package:pubspec_parse/pubspec_parse.dart' hide Pubspec;
+
+const indentedEncoder = JsonEncoder.withIndent('  ');
 
-  AnalysisContext contextFor(String path) =>
-      analysisContextCollection.contextFor(path);
+abstract class PackageAnalysisContext {
+  AnalysisSession get analysisSession;
 
-  /// Log [message] as warning that something unexpected happened.
+  /// Log [message] as a warning that something unexpected happened.
   void warning(String message);
+
+  /// Get the contents of the file at [path] using the `analysisSession.resourceProvider`
+  String readFile(String path) =>
+      analysisSession.resourceProvider.getFile(path).readAsStringSync();
+
+  resource.File getFile(String path) =>
+      analysisSession.resourceProvider.getFile(path);
+
+  resource.Folder getFolder(String path) =>
+      analysisSession.resourceProvider.getFolder(path);
+
+  String? uriToPath(Uri uri) => analysisSession.uriConverter.uriToPath(uri);
+}
+
+/// Download version [version] of the package [name] to the directory
+/// [destination] and fetch its dependencies, optionally deleting the contents
+/// of [destination] first if [wipeTarget] is set to true.
+///
+/// The package will be downloaded using `downloadPackage` from
+/// `package:pana/src/download_utils.dart`.
+Future<void> fetchPackageAndDependencies({
+  required String name,
+  required String version,
+  required String destination,
+  required bool wipeTarget,
+}) async {
+  // delete the target directory, if it exists and wipe is enabled
+  if (wipeTarget && await Directory(destination).exists()) {
+    await Directory(destination).delete(recursive: true);
+  }
+
+  // download the package and fetch dependencies
+  await downloadPackage(name, version, destination: destination);
+  await fetchDependencies(destination);
+  return;
+}
+
+/// Download version [version] of the package [name] to the directory
+/// [destination] and fetch its dependencies, optionally deleting the contents
+/// of [destination] first if [wipeTarget] is set to true.
+///
+/// A pointer package will be created at [destination] with a single dependency,
+/// pinning the dependency [name] to version [version].
+Future<void> fetchPackageWithPointer({
+  required String name,
+  required String version,
+  required String destination,
+  required bool wipeTarget,
+}) async {
+  // delete the target directory, if it exists and wipe is enabled
+  if (wipeTarget && await Directory(destination).exists()) {
+    await Directory(destination).delete(recursive: true);
+  }
+
+  // construct pubspec for the pointer package
+  final pointerPubspec = <String, dynamic>{
+    'name': 'dummy_package',
```

Comment:
Consider using dot in the package name:
```suggestion
    'name': 'dummy.package',
```

That way it can't be published, so it can't conflict with a package that exists on pub.dev :D